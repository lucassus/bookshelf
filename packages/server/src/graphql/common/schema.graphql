scalar ExternalID

interface Resource {
  id: ExternalID!
}

union Anything = Author | Book | User | BookCopy

interface Timestampable {
  createdAt: String!
  updatedAt: String!
}

type Image {
  path: String!
  url: String!
}

extend type Query {
  resource(id: ID!): Resource!
  anything(id: ID!): Anything! @deprecated(reason: "No longer supported. Use 'resources' instead")
}
