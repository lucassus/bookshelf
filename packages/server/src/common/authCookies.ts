import express from "express";

import {
  AUTH_COOKIE_NAME,
  AUTH_TOKEN_EXPIRES_IN_SECONDS,
  Environment
} from "../config";

export const sendAuthCookie = (res: express.Response, authToken: string) =>
  res.cookie(AUTH_COOKIE_NAME, authToken, {
    httpOnly: true,
    secure: process.env.NODE_ENV === Environment.production,
    maxAge: AUTH_TOKEN_EXPIRES_IN_SECONDS * 1000
  });

export const clearAuthCookie = (res: express.Response) =>
  res.clearCookie(AUTH_COOKIE_NAME);
