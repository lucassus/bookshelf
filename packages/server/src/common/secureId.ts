import { SecureId } from "@bookshelf/secure-id";

export const secureId = new SecureId<string>({
  separator: "-"
});
