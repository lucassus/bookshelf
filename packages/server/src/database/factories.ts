import { getManager } from "typeorm";

import { Author } from "./entity/Author";
import { Avatar } from "./entity/Avatar";
import { Book } from "./entity/Book";
import { BookCopy } from "./entity/BookCopy";
import { User } from "./entity/User";

export const createAvatar = (attributes: Partial<Avatar> = {}) => {
  const manager = getManager();

  return manager.save(
    manager.create(Avatar, {
      imagePath: "/images/avatars/w13.png",
      color: "yellow",
      ...attributes
    })
  );
};

export const createUser = async (
  attributes: Partial<User> = { name: "Alice" }
) => {
  const manager = getManager();

  const { ...userAttributes } = attributes;

  if (userAttributes.name && userAttributes.email === undefined) {
    userAttributes.email = `${userAttributes.name.toLocaleLowerCase()}@email.com`;
  }

  if (userAttributes.avatarId === undefined) {
    const avatar = await createAvatar();
    userAttributes.avatarId = avatar.id;
  }

  const user = manager.create(User, userAttributes);

  return manager.save(user);
};

export const createAuthor = (attributes: Partial<Author> = {}) => {
  const manager = getManager();

  return manager.save(
    manager.create(Author, {
      name: "Andrzej Sapkowski",
      photoPath: "/images/book-authors/andrzej-sapkowski.jpg",
      ...attributes
    })
  );
};

export const createBook = async (
  attributes: Partial<Book> & { authorAttributes?: Partial<Author> } = {}
) => {
  const manager = getManager();

  const { authorAttributes, ...bookAttributes } = attributes;

  if (bookAttributes.authorId === undefined) {
    const author = await createAuthor(authorAttributes);
    bookAttributes.authorId = author.id;
  }

  return manager.save(
    manager.create(Book, {
      title: "Baptism of fire",
      coverPath: "/images/book-covers/witcher3.jpg",
      ...bookAttributes
    })
  );
};

export const createBookCopy = async (
  attributes: Partial<BookCopy> & { bookAttributes?: Partial<Book> } & {
    ownerAttributes?: Partial<User>;
  } = {}
) => {
  const manager = getManager();

  const { bookAttributes, ownerAttributes, ...bookCopyAttributes } = attributes;

  if (bookCopyAttributes.bookId === undefined) {
    const book = await createBook(bookAttributes);
    bookCopyAttributes.bookId = book.id;
  }

  if (bookCopyAttributes.ownerId === undefined) {
    const user = await createUser(ownerAttributes);
    bookCopyAttributes.ownerId = user.id;
  }

  return manager.save(manager.create(BookCopy, bookCopyAttributes));
};
