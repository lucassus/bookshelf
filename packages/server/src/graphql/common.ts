import { Author } from "../database/entity/Author";
import { Avatar } from "../database/entity/Avatar";
import { Book } from "../database/entity/Book";
import { Context } from "../types";
import { Resolver, ResolversTypes } from "./resolvers-types.generated";

// TODO: Figure out how to refactor it with interface
export const image: Resolver<
  ResolversTypes["Image"],
  Avatar | Author | Book,
  Context
> = (rootValue, args, { assetsBaseUrl }) => {
  let path = "/default.jpg";

  if (rootValue instanceof Avatar) {
    path = rootValue.imagePath;
  }

  if (rootValue instanceof Author) {
    path = rootValue.photoPath;
  }

  if (rootValue instanceof Book) {
    path = rootValue.coverPath;
  }

  return {
    path,
    url: assetsBaseUrl + path
  };
};
