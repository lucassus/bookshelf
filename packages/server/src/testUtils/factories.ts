import faker from "faker";
import { getManager, ObjectType, DeepPartial } from "typeorm";

import { hashPassword } from "../common/authentication";
import { titleizeSentence } from "../common/strings";
import { Author } from "../database/entity/Author";
import { Avatar, AVATAR_COLORS } from "../database/entity/Avatar";
import { Book } from "../database/entity/Book";
import { BookCopy } from "../database/entity/BookCopy";
import { User } from "../database/entity/User";

const AVATAR_IMAGES = [
  "/images/avatars/w13.png",
  "/images/avatars/m10.png",
  "/images/avatars/w2.png",
  "/images/avatars/m25.png"
];

const AUTHOR_PHOTOS = [
  "/images/book-authors/j-k-rowling.jpg",
  "/images/book-authors/james-s-a-corey.jpg",
  "/images/book-authors/andrzej-sapkowski.jpg"
];

const BOOK_COVERS = [
  "/images/book-covers/harry1.jpg",
  "/images/book-covers/harry2.jpg",
  "/images/book-covers/harry3.jpg",
  "/images/book-covers/harry4.jpg",
  "/images/book-covers/harry5.jpg",
  "/images/book-covers/harry6.jpg",
  "/images/book-covers/harry7.jpg",
  "/images/book-covers/expanse1.jpg",
  "/images/book-covers/expanse2.jpg",
  "/images/book-covers/expanse3.jpg",
  "/images/book-covers/expanse4.jpg",
  "/images/book-covers/expanse5.jpg",
  "/images/book-covers/expanse6.jpg",
  "/images/book-covers/expanse7.jpg",
  "/images/book-covers/expanse8.jpg",
  "/images/book-covers/witcher1.jpg",
  "/images/book-covers/witcher2.jpg",
  "/images/book-covers/witcher3.jpg",
  "/images/book-covers/witcher4.jpg",
  "/images/book-covers/witcher5.jpg"
];

function createEntity<Entity>(
  entityClass: ObjectType<Entity>,
  attributes: DeepPartial<Entity>
) {
  const manager = getManager();

  const entity = manager.create(entityClass, attributes);
  return manager.save(entity);
}

type CreateAvatarAttributes = Partial<Avatar>;

export function createAvatar(attributes: CreateAvatarAttributes = {}) {
  return createEntity(Avatar, {
    imagePath: faker.random.arrayElement(AVATAR_IMAGES),
    color: faker.random.arrayElement(AVATAR_COLORS),
    ...attributes
  });
}

type CreateUserAttributes = Partial<User> & {
  password?: string;
  avatarAttributes?: CreateAvatarAttributes;
};

export async function createUser(attributes: CreateUserAttributes = {}) {
  const { avatarAttributes, password, ...userAttributes } = attributes;

  if (userAttributes.name === undefined) {
    userAttributes.name = faker.name.findName();
  }

  if (userAttributes.email === undefined) {
    userAttributes.email = faker.internet.email();
  }

  if (userAttributes.info === undefined) {
    userAttributes.info = faker.lorem.sentence();
  }

  if (userAttributes.avatarId === undefined) {
    const avatar = await createAvatar();
    userAttributes.avatarId = avatar.id;
  }

  if (avatarAttributes) {
    const avatar = await createAvatar(avatarAttributes);
    userAttributes.avatarId = avatar.id;
  }

  return createEntity(User, {
    passwordHash: hashPassword(password || "password"),
    isAdmin: false,
    ...userAttributes
  });
}

export function createAuthor(attributes: Partial<Author> = {}) {
  return createEntity(Author, {
    name: faker.name.findName(),
    bio: faker.lorem.sentence(),
    photoPath: faker.random.arrayElement(AUTHOR_PHOTOS),
    ...attributes
  });
}

type CreateBookAttributes = Partial<Book> & {
  authorAttributes?: Partial<Author>;
};

export async function createBook(attributes: CreateBookAttributes = {}) {
  const { authorAttributes, ...bookAttributes } = attributes;

  if (bookAttributes.authorId === undefined) {
    const author = await createAuthor(authorAttributes);
    bookAttributes.authorId = author.id;
  }

  return createEntity(Book, {
    title: titleizeSentence(
      faker.lorem.words(faker.random.number({ min: 1, max: 4 }))
    ),
    description: faker.lorem.sentence(),
    coverPath: faker.random.arrayElement(BOOK_COVERS),
    favourite: faker.random.boolean(),
    ...bookAttributes
  });
}

export async function createBookCopy(
  attributes: Partial<BookCopy> & {
    bookAttributes?: CreateBookAttributes;
  } & {
    ownerAttributes?: CreateUserAttributes;
  } & { borrowerAttributes?: CreateUserAttributes } = {}
) {
  const {
    bookAttributes,
    ownerAttributes,
    borrowerAttributes,
    ...bookCopyAttributes
  } = attributes;

  if (bookCopyAttributes.bookId === undefined) {
    const book = await createBook(bookAttributes);
    bookCopyAttributes.bookId = book.id;
  }

  if (bookCopyAttributes.ownerId === undefined) {
    const owner = await createUser(ownerAttributes);
    bookCopyAttributes.ownerId = owner.id;
  }

  if (borrowerAttributes) {
    const borrower = await createUser(borrowerAttributes);
    bookCopyAttributes.borrowerId = borrower.id;
  }

  return createEntity(BookCopy, bookCopyAttributes);
}
