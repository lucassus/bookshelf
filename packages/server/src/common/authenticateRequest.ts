import express from "express";
import jwt from "jsonwebtoken";
import { getRepository } from "typeorm";

import { AUTH_COOKIE_NAME } from "../config";
import { User } from "../database/entity";
import { getAuthTokenSecretFor } from "./authTokens";

export async function authenticateRequest(
  req: express.Request
): Promise<undefined | User> {
  const { [AUTH_COOKIE_NAME]: authToken } = req.cookies;
  if (!authToken) {
    return undefined;
  }

  const payload = jwt.decode(authToken);
  if (!(payload && payload.sub)) {
    throw new Error("Invalid token payload");
  }

  const user = await getRepository(User).findOneOrFail({
    id: payload.sub
  });

  return new Promise((resolve, reject) => {
    jwt.verify(authToken, getAuthTokenSecretFor(user), (error) => {
      if (error) {
        reject(error);
      } else {
        resolve(user);
      }
    });
  });
}
