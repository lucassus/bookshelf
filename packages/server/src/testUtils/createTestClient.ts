import { ApolloServer } from "apollo-server-express";
import {
  ApolloServerTestClient,
  createTestClient as createApolloTestClient
} from "apollo-server-testing";

import { buildContext, Context } from "../graphql/context";
import { rootSchema } from "../graphql/rootSchema";

export function createTestClient(
  contextExtra: Partial<Context> = {}
): ApolloServerTestClient {
  const server = new ApolloServer({
    schema: rootSchema,
    context: buildContext(contextExtra)
  });

  return createApolloTestClient(server);
}
