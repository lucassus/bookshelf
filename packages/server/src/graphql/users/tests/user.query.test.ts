import { gql } from "apollo-server-express";

import { secureId } from "../../../common/secureId";
import { createTestClient } from "../../../testUtils/createTestClient";
import { createBookCopy, createUser } from "../../../testUtils/factories";

test("user query", async () => {
  // Given
  const user = await createUser();
  await createBookCopy({ owner: user });
  await createBookCopy({ owner: user });
  await createBookCopy({ borrower: user });

  // When
  const res = await createTestClient().query({
    query: gql`
      query($id: ExternalID!) {
        user(id: $id) {
          name
          info
          avatar {
            color
            image {
              url
            }
          }
          ownedBookCopies {
            book {
              id
              title
            }
            owner {
              id
              name
            }
          }
          borrowedBookCopies {
            borrower {
              id
              name
            }
            book {
              id
              title
            }
          }
        }
      }
    `,
    variables: { id: secureId.toExternal(user.id, "User") }
  });

  // Then
  expect(res.errors).toBe(undefined);
  expect(res.data).not.toBeNull();
  expect(res.data!.user).toMatchSnapshot();
});
