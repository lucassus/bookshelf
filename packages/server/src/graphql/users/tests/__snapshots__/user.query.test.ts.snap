// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`user query 1`] = `
Object {
  "avatar": Object {
    "color": "red",
    "image": Object {
      "url": "https://res.cloudinary.com/lucassus/image/upload/bookshelf/images/avatars/w2.png",
    },
  },
  "borrowedBookCopies": Array [
    Object {
      "book": Object {
        "id": "My1Cb29r",
        "title": "Alias",
      },
      "borrower": Object {
        "id": "MS1Vc2Vy",
        "name": "Mariana Christiansen",
      },
    },
  ],
  "info": "Velit quaerat ab veniam.",
  "name": "Mariana Christiansen",
  "ownedBookCopies": Array [
    Object {
      "book": Object {
        "id": "MS1Cb29r",
        "title": "Minus",
      },
      "owner": Object {
        "id": "MS1Vc2Vy",
        "name": "Mariana Christiansen",
      },
    },
    Object {
      "book": Object {
        "id": "Mi1Cb29r",
        "title": "Earum Quaerat Architecto",
      },
      "owner": Object {
        "id": "MS1Vc2Vy",
        "name": "Mariana Christiansen",
      },
    },
  ],
}
`;
